src/credentials/CertificationDeclaration.cpp

```diff
@@ -245,57 +283,74 @@ CHIP_ERROR DecodeCertificationElements(const ByteSpan & encodedCertElements, Cer
 
 bool CertificationElementsDecoder::IsProductIdIn(const ByteSpan & encodedCertElements, uint16_t productId)
 {
-    VerifyOrReturnError(PrepareToReadProductIdList(encodedCertElements) == CHIP_NO_ERROR, false);
+    VerifyOrReturnError(FindAndEnterArray(encodedCertElements, ContextTag(kTag_ProductIdArray)) == CHIP_NO_ERROR, false);
 
     uint16_t cdProductId = 0;
-    CHIP_ERROR error     = CHIP_NO_ERROR;
-
-    while ((error = GetNextProductId(cdProductId)) == CHIP_NO_ERROR)
+    while (GetNextProductId(cdProductId) == CHIP_NO_ERROR)
     {
         if (productId == cdProductId)
         {
             return true;
         }
     }
-
     return false;
 }
 
-CHIP_ERROR CertificationElementsDecoder::PrepareToReadProductIdList(const ByteSpan & encodedCertElements)
+CHIP_ERROR CertificationElementsDecoder::FindAndEnterArray(const ByteSpan & encodedCertElements, Tag arrayTag)
 {
-    mIsInitialized                = false;
-    mCertificationDeclarationData = encodedCertElements;
+    TLVType outerContainerType1;
+    TLVType outerContainerType2;
 
-    mReader.Init(mCertificationDeclarationData);
+    mReader.Init(encodedCertElements);
     ReturnErrorOnFailure(mReader.Next(kTLVType_Structure, AnonymousTag()));
-    ReturnErrorOnFailure(mReader.EnterContainer(mOuterContainerType1));
+    ReturnErrorOnFailure(mReader.EnterContainer(outerContainerType1));
 
     // position to ProductId Array
     CHIP_ERROR error = CHIP_NO_ERROR;
     do
     {
-        error = mReader.Next(kTLVType_Array, ContextTag(kTag_ProductIdArray));
+        error = mReader.Next(kTLVType_Array, arrayTag);
         // return error code if Next method returned different than CHIP_NO_ERROR.
         // also return if different error code than CHIP_ERROR_WRONG_TLV_TYPE/CHIP_ERROR_UNEXPECTED_TLV_ELEMENT, which means that
         // the expected type and tags do not match.
```

Comment:
Pre-existing, but this should say:
```suggestion
        // Return error code unless one of three things happened:
        // 1. We found the right thing (CHIP_NO_ERROR returned).
        // 2. The next tag is not the one we are looking for (CHIP_ERROR_UNEXPECTED_TLV_ELEMENT).
        // 3. The next thing has the right tag but the wrong type (CHIP_ERROR_WRONG_TLV_TYPE).
```
But when I write it like that ... why is case 3 not an error condition?